onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorRotaryEmbedding.cpp

```diff
@@ -0,0 +1,412 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "precomp.h"
+
+// This operator is easier to understand by looking at a python implementation of the non-interleaved version:
+//
+// def rotate_half(x):
+//     """Rotates half the hidden dims of the input."""
+//     half_dim = x.shape[-1] // 2
+//     x1 = x[..., :half_dim]
+//     x2 = x[..., half_dim:]
+//     return np.concatenate((-x2, x1), dim=-1)
+//
+//
+// def apply_rope(x, cos, sin, position_ids):
+//     cos = cos[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
+//     sin = sin[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
+//     x_embed = (x * cos) + (rotate_half(x) * sin)
+//     return x_embed
+//
+// For the non-interleaved version, we multiply the cos cache by the non-rotated input tensor while we multiply the sin cache
+// by the rotated input tensor. Rotating the tensor means slicing it in half on the head dimension and swapping the 2 halves.
+//
+// The interleaved version is very similar but instead of swapping 2 halves, we swap every pair of adjacent elements and we swap
+// the sign of every adjacent element.
+
+namespace Dml
+{
+class DmlOperatorRotaryEmbedding : public DmlOperator
+{
+public:
+    DmlOperatorRotaryEmbedding(const MLOperatorKernelCreationContext& kernelInfo) : DmlOperator(kernelInfo)
+    {
+        enum InputIndex : uint32_t
+        {
+            inputDataIndex,
+            positionIdsIndex,
+            cosCacheIndex,
+            sinCacheIndex,
+        };
+
+        ML_CHECK_VALID_ARGUMENT(kernelInfo.GetInputCount() == 4);
+        ML_CHECK_VALID_ARGUMENT(kernelInfo.GetOutputCount() == 1);
+
+        // When positionIds is a scalar, it represents the start offset for each sequence
+        const bool positionIdsIsOffset = kernelInfo.GetInputTensorDimensionCount(positionIdsIndex) == 1;
+
+        Initialize(kernelInfo);
+
+        ML_CHECK_VALID_ARGUMENT(m_inputTensorDescs[inputDataIndex].GetDimensionCount() == 4);
+        ML_CHECK_VALID_ARGUMENT(m_inputTensorDescs[positionIdsIndex].GetDimensionCount() == 4);
+        ML_CHECK_VALID_ARGUMENT(m_inputTensorDescs[cosCacheIndex].GetDimensionCount() == 4);
+        ML_CHECK_VALID_ARGUMENT(m_inputTensorDescs[sinCacheIndex].GetDimensionCount() == 4);
+
+        ML_CHECK_VALID_ARGUMENT(m_outputTensorDescs[0].GetDimensionCount() == 4);
+
+        ML_CHECK_VALID_ARGUMENT(m_inputTensorDescs[cosCacheIndex].GetSizes() == m_inputTensorDescs[sinCacheIndex].GetSizes());
+        const uint32_t headSize = m_inputTensorDescs[cosCacheIndex].GetSizes().back() * 2;
```

Comment:
If I am not mistaken, gsl::span == operator compares the pointer not the actual values.
[https://github.com/microsoft/GSL/blob/e64c97fc2cfc11992098bb38eda932de275e3f4d/include/gsl/span#L247](https://github.com/microsoft/GSL/blob/e64c97fc2cfc11992098bb38eda932de275e3f4d/include/gsl/span#L247)

If that's what intended here, then I think `cosCache` and `sinCache` will not have same gsl::span object.

Reply:
You're looking at the `span_iterator` comparison. The `span` comparison is over here: https://github.com/microsoft/GSL/blob/e64c97fc2cfc11992098bb38eda932de275e3f4d/include/gsl/span_ext#L57C8-L57C8

Reply:
Ahh I see.